### PR TITLE
#149956358 Update contributions endpoints

### PR DESCRIPTION
**Pull request by andela-milesanmi**
#### What does this PR do?
- Update contribution endpoints to contain all data coming from and required by the frontend
#### Description of Task to be completed?
- Implement the saving and fetching of Tags
- Update all existing endpoints to use the data from the frontend
- Extract database saving functionality into an inherited function
- Update tests for the endpoints
#### How should this be manually tested?
At the root of the project directory run the following command:
`DB_TYPE=sqlite pytest` or `DB_TYPE=sqlite pytest -v`
#### What are the relevant pivotal tracker stories?
#149956358
#### Screenshots
- Sample response
screen shot 2017-08-09 at 21 03 54

### CODE REVIEW

File: api/views/contributions.py [hunk 1]
```diff
@@ -108,21 +108,34 @@ def post(self):
...
-            db.session.commit()
+            new_contribution.save()
 
             contribution_serialized = new_contribution.serialize()
 
             if "tags" in result:
                 for contribution_tag in result["tags"]:
-                    valid_tag = Tag.query.filter_by(id=contribution_tag).first()
+                    valid_tag = Tag.query.filter_by(name=contribution_tag).first()
                     if not valid_tag:
```
Comment (andela-oowonikoko) [inline api/views/contributions.py]: I don't know if this would be possible, so feel free to correct me if it isn't. The code you have on line 117-131 and 171-192 (that saves tags into the db and append it to the response) are similar. So I feel you should extract it into the helper.py file, so you can import it instead of repeating it 😃

Comment (andela-milesanmi) [inline api/views/contributions.py]: Okay. I believe that is a valid point. Will work to streamline it. Thanks

File: api/models.py [hunk 2]
```diff
@@ -14,7 +40,29 @@ class SerializerMixin(object):
     def serialize(self):
         return {column.name: getattr(self, column.name)
                 for column in self.__table__.columns
-                if column.name not in ['contribution_type_id']}
+                if column.name not in [ 'contribution_type_id' ]}
+
+    def save(self):
+        db.session.add(self)
+        db.session.commit()
```
Comment (andela-uibrahim) [inline api/models.py]: nice one

File: api/models.py [hunk 3]
```diff
@@ -14,7 +40,29 @@ class SerializerMixin(object):
...
+    #             # Dynamically create tags.
+    #             generated_tag_id = '-K' + str(uuid4().hex).upper()[0:6]
+    #             new_tag = Tag(
+    #                 id=generated_tag_id,
+    #                 name=contribution_tag
+    #             )
+    #             new_tag.save()
+    #             contribution.tags.append(new_tag)
+    #         else:
+    #             contribution.tags.extend(valid_tag)
```
Comment (andela-uibrahim) [inline api/models.py]: do not push commented out code

File: api/models.py [hunk 4]
```diff
@@ -46,6 +94,9 @@ class Tag(db.Model, SerializerMixin):
 
     id = db.Column(db.String, primary_key=True)
     name = db.Column(db.String(80), unique=True)
+    contributions = db.relationship('Contribution',
+                         secondary="contribution_tag",
+                         back_populates='tags')
```
Comment (andela-uibrahim) [inline api/models.py]: why did you do this?

File: api/views/contributions.py [hunk 5]
```diff
@@ -84,14 +84,14 @@ def post(self):
 
         if request.get_json():
             result = request.get_json()
-            keys = ["title", "description", "tags", "type"]
+            keys = ["title", "description", "tags", "type", "meta"]
 
             for key in keys:
-                if key is not "tags" and key not in result:
+                if key not in ["tags"] and key not in result:
```
Comment (andela-uibrahim) [inline api/views/contributions.py]: what were you trying to achieve here?

File: api/views/contributions.py [hunk 6]
```diff
@@ -84,14 +84,14 @@ def post(self):
...
                     return {"message": key + " cannot be null"}, 400
                 elif (key in ["title", "description", "type"]) and not validate_type(result[key], str):
                     return {"message": key + " must be a valid string"}, 400
-                elif (key is "tags" and key in result) and not validate_type(result[key], list):
+                elif (key in ["tags"] and key in result) and not validate_type(result[key], list):
```
Comment (andela-uibrahim) [inline api/views/contributions.py]: what were you trying to achieve here?

File: api/views/contributions.py [hunk 7]
```diff
@@ -146,31 +159,53 @@ def put(self, contrib_id):
...
+                    db.session.commit()
+                    
+                    for contribution_tag in data["tags"]:
+                        valid_tag = Tag.query.filter_by(name=contribution_tag).all()
+                        if not valid_tag:
+                            # Dynamically create tags.
+                            generated_tag_id = '-K' + str(uuid4().hex).upper()[0:6]
+                            new_tag = Tag(
+                                id=generated_tag_id,
+                                name=contribution_tag
```
Comment (andela-uibrahim) [inline api/views/contributions.py]: we need to check with @njirap if she supports this idea

File: api/views/contributions.py [hunk 8]
```diff
@@ -108,21 +108,34 @@ def post(self):
...
             contribution_serialized = new_contribution.serialize()
 
             if "tags" in result:
                 for contribution_tag in result["tags"]:
-                    valid_tag = Tag.query.filter_by(id=contribution_tag).first()
+                    valid_tag = Tag.query.filter_by(name=contribution_tag).first()
                     if not valid_tag:
-                        return {"message": "{} is not a valid tag".format(contribution_tag)}, 400
-                    new_contribution.tags.append(valid_tag)
+                        generated_tag_id = '-K' + str(uuid4().hex).upper()[0:6]
```
Comment (andela-uibrahim) [inline api/views/contributions.py]: nice idea. also the ids can be  made longer and njira said we should generate this in the model as opposed to the view

**Comment by andela-uibrahim** [comment]
please squash the commits to 1

File: api/models.py (hunk 3 above)
Comment (njirap) [inline api/models.py]: What Usman said

File: api/views/contributions.py (hunk 8 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: I created the ids that way since we dont have an exact id length. I believe once the task open on Pivotal tracker to auto-create `ids` is completed, this issue will be resolved

File: api/views/contributions.py (hunk 7 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: which idea?

File: api/models.py (hunk 2 above)
Comment (andela-milesanmi) [inline api/models.py]: thank you

File: api/models.py (hunk 4 above)
Comment (andela-milesanmi) [inline api/models.py]: This is necessary to update the `contribution_tag` table with the contribution and tag ids based on association.

File: api/helpers.py [hunk 9]
```diff
@@ -1,2 +1,25 @@
...
+
 def validate_type(item, input_type):
     return type(item) is input_type
+
+
+def save_tag(data, contribution):
+
+    # contribution.tags = []
+    for contribution_tag in data["tags"]:
+        valid_tag = Tag.query.filter_by(name=contribution_tag).first()
```
Comment (njirap) [inline api/helpers.py]: tags from the front end are coming in by `id` and not by `name`

File: api/helpers.py [hunk 10]
```diff
@@ -1,2 +1,25 @@
...
+            # Dynamically create tags.
+            generated_tag_id = '-K' + str(uuid4().hex).upper()[0:6]
+            new_tag = Tag(
+                id=generated_tag_id,
+                name=contribution_tag
+            )
+            new_tag.save()
+            contribution.tags.append(new_tag)
+        else:
+            contribution.tags.append(valid_tag)
```
Comment (njirap) [inline api/helpers.py]: 1. remove L21 and 22
2. change `new_tag` and `valid_tag` var names to `tag`
3. unindent L23 to be at same level with the `if on L13.

This will remove the repetition of Lines 21 and 23

File: api/helpers.py [hunk 11]
```diff
@@ -1,2 +1,25 @@
+from .models import Tag
+from uuid import uuid4
+
 def validate_type(item, input_type):
     return type(item) is input_type
+
+
+def save_tag(data, contribution):
+
+    # contribution.tags = []
```
Comment (njirap) [inline api/helpers.py]: remove commented code

File: api/helpers.py (hunk 9 above)
Comment (andela-milesanmi) [inline api/helpers.py]: I was told that the tags will be sent as an array of strings. This is to enable us auto-generate tags, as was requested by kes.

File: api/helpers.py (hunk 11 above)
Comment (andela-milesanmi) [inline api/helpers.py]: done.

File: api/helpers.py (hunk 10 above)
Comment (andela-milesanmi) [inline api/helpers.py]: Okay. Done

File: api/models.py (hunk 2 above)
Comment (njirap) [inline api/models.py]: yeah very great! thanks!

Comment (njirap) [inline api/models.py]: the mixin then ceases to be just a `SerializerMixin` because i does more than just serialize, it now saves also. So maybe a more suitable name would be `ModelOpsMixin`

File: api/models.py [hunk 12]
```diff
@@ -14,7 +39,14 @@ class SerializerMixin(object):
...
                 for column in self.__table__.columns
-                if column.name not in ['contribution_type_id']}
+                if column.name not in [ 'contribution_type_id' ]}
+
+    def save(self):
+        db.session.add(self)
+        db.session.commit()
+
+    def delete(self):
+        pass
```
Comment (njirap) [inline api/models.py]: why is this here?

File: api/models.py (hunk 2 above)
Comment (andela-milesanmi) [inline api/models.py]: thank you. Fixed that.

File: api/models.py (hunk 12 above)
Comment (andela-milesanmi) [inline api/models.py]: this is meant to be the method to handle delete operations. I didn't fill it out since we haven't started working on delete operations yet.

File: api/views/contributions.py [hunk 13]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
-        else:
+        if not contribution:
             response = jsonify(dict(status=404,
-                                    message='Contribution does not exist'))
+                                message='Contribution does not exist'))
             response.status_code = 404
             return response
+        
+        if "contribution_type_id" in data:
+            get_contribution_type = ContributionType.query.get(data["contribution_type_id"])
```
Comment (njirap) [inline api/views/contributions.py]: great catch!

File: api/views/contributions.py [hunk 14]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
+            if not get_contribution_type:
+                response = jsonify(dict(
+                    message='Invalid contribution type'
+                ))
+                response.status_code = 404
+                return response
+            contribution.contribution_type.name = ContributionType.query.get(data["contribution_type_id"]).name
```
Comment (njirap) [inline api/views/contributions.py]: you have already queried for the ContributionType and assigned it a variable `get_contribution_type` no need to query again.

Also please rename the variable `get_contribution_type` to something like `contrib_type`

File: api/views/contributions.py [hunk 15]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
+        
+        contribution_serialized = contribution.serialize()
+        for key in data:
+            if "tags" in data:
+                # Remove deleted tags from the Association table
+                contribution.tags = []
+                db.session.commit()
```
Comment (njirap) [inline api/views/contributions.py]: does something have to be added to the db session before it can be commited?

Comment (njirap) [inline api/views/contributions.py]: also what and why are you commiting here?

File: api/helpers.py [hunk 16]
```diff
@@ -1,2 +1,25 @@
+from .models import Tag
+from uuid import uuid4
+
 def validate_type(item, input_type):
     return type(item) is input_type
+
+
+def save_tag(data, contribution):
```
Comment (njirap) [inline api/helpers.py]: change the name to `save_tags` to reflect that it is not saving 1 tag but many

File: api/views/contributions.py (hunk 14 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: noted. done

File: api/views/contributions.py (hunk 13 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: thanks

File: api/helpers.py (hunk 16 above)
Comment (njirap) [inline api/helpers.py]: since the saving of the tags is specific to the `Contribution` models I suggest we have this function in the model itself

File: api/views/contributions.py [hunk 17]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
+
+                tagged_contribution = save_tag(data, contribution)
+
+                contribution_serialized['tags'] = [tag.serialize()
+                                        for tag in tagged_contribution.tags]
```
Comment (njirap) [inline api/views/contributions.py]: this will not be necessary when the saving of the tags becomes a model function in `Contribution`

File: api/helpers.py (hunk 16 above)
Comment (andela-milesanmi) [inline api/helpers.py]: Okay

File: api/views/contributions.py [hunk 18]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
+            else:
+                tagged_contribution = contribution
+
+            if key not in ['id', 'user_id', 'tags', 'type']:
+                setattr(tagged_contribution, key, data[key])
```
Comment (njirap) [inline api/views/contributions.py]: why are we saving unplanned for fields?

File: api/views/contributions.py [hunk 19]
```diff
@@ -146,31 +145,54 @@ def put(self, contrib_id):
...
+        
+        # Save the updated contribution to the database
+        tagged_contribution.save()
+
+        # add contribution_type field
+        contribution_serialized['contribution_type'] = contribution.contribution_type.name
+        contribution_serialized['status'] = tagged_contribution.status
```
Comment (njirap) [inline api/views/contributions.py]: if you edit your logic as above the use of 2 contributions will not be necessary

File: api/views/contributions.py (hunk 17 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: okay @njirap. but taking it out now would mean in inability to save tags. I believe it cant be left till the task is picked up and worked on from here.

File: test/api/views/test_contributions.py [hunk 20]
```diff
@@ -351,6 +357,7 @@ def test_tags_is_list(self):
         self.assertEqual(res.status_code, 400)
         self.assertEqual(expected_response, generated_response['message'])
 
+    @pytest.mark.skip(reason="The tags are now being created on the fly")
```
Comment (njirap) [inline test/api/views/test_contributions.py]: it then means that the test is null and void hence instead of skipping it just delete it

File: test/api/views/test_contributions.py [hunk 21]
```diff
@@ -444,8 +458,9 @@ def test_update_status(self):
             id=contrib_id,
             title="AI",
             user_id="-KJKcv0D",
-            contribution_type_id="-KSDQcv0D",
+            contribution_type_id="-KSDQcv0D"
         )
+        contrib.contribution_type = self.resource
```
Comment (njirap) [inline test/api/views/test_contributions.py]: what is the purpose of this?

File: api/views/contributions.py (hunk 18 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: do you mean that we should check for what can be updated instead of what can't be updated?

File: test/api/views/test_contributions.py (hunk 20 above)
Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: okay. i wanted to ensure that you were onboard with creating tags dynamically before taking it out.

File: test/api/views/test_contributions.py [hunk 22]
```diff
@@ -538,6 +553,40 @@ def test_update_contribution_type(self):
             data=json.dumps(data),
             content_type='application/json'
         )
+        new_contrib_contribution_type_id = Contribution.query.filter_by(id=contrib_id).first().contribution_type_id
```
Comment (njirap) [inline test/api/views/test_contributions.py]: what are you using this variable for?

File: test/api/views/test_contributions.py [hunk 23]
```diff
@@ -538,6 +553,40 @@ def test_update_contribution_type(self):
...
+    
+    def test_update_contribution_type(self):
+        contrib_id = "-KDfy7648W"
+
+        idea = ContributionType(
+            id = "-KSDQcv0D",
+            name="resource",
+        )
+        db.session.add(idea)
+        db.session.commit()
```
Comment (njirap) [inline test/api/views/test_contributions.py]: don't we already have `idea.save`?

File: test/api/views/test_contributions.py [hunk 24]
```diff
@@ -538,6 +553,40 @@ def test_update_contribution_type(self):
...
+
+        contrib = Contribution(
+            id=contrib_id,
+            title="Javascript",
+            user_id="-KJKcv0D",
+            description="Lorem ipsum dolor sit amet.",
+            contribution_type_id=idea.id,
+        )
+        db.session.add(contrib)
+        db.session.commit()
```
Comment (njirap) [inline test/api/views/test_contributions.py]: don't we already have `contrib.save`

Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: yes. those have been effected.

File: test/api/views/test_contributions.py (hunk 22 above)
Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: this is used in asserting that the contribution is actually different from what it was originally

File: test/api/views/test_contributions.py [hunk 25]
```diff
@@ -538,6 +553,40 @@ def test_update_contribution_type(self):
...
+        old_contrib_contribution_type_id = Contribution.query.filter_by(id=contrib_id).first().contribution_type_id
```
Comment (njirap) [inline test/api/views/test_contributions.py]: let's not exaggerate on the variable names.. we don't want variable sentences.. for this something like `old_contrib_type` makes sense and is shorter

Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: sawa

File: test/api/views/test_contributions.py (hunk 22 above)
Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: this was unnecessary. fixed. thanks

File: test/api/views/test_contributions.py (hunk 21 above)
Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: it adds the `resource` contribution info to the contribution

File: api/views/contributions.py (hunk 18 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: corrected..

File: api/views/contributions.py (hunk 15 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: the commit was not needed. it was a fragment of the initial implementation i tried.

**Comment by andela-milesanmi** [comment]
Hi. @njirap. All the issues raised have been fixed. Thanks

File: api/models.py [hunk 26]
```diff
@@ -1,21 +1,49 @@
 from datetime import datetime
-
 from flask_sqlalchemy import SQLAlchemy
+from sqlalchemy.orm import backref
+from sqlalchemy.dialects.postgresql.json import JSONB
```
Comment (njirap) [inline api/models.py]: @andela-milesanmi this forces us all to use postgres. please make it engine agnostic

Comment (andela-milesanmi) [inline api/models.py]: @njirap, I'm implementing it like this because postgres is what we're currently using on development and production. This is the only way I could think of.

Comment (andela-milesanmi) [inline api/models.py]: @njirap, done. thanks

File: api/models.py (hunk 4 above)
Comment (LarryWachira) [inline api/models.py]: @andela-milesanmi This had already been done. No refactoring was necessary.

**Comment by njirap** [comment]
LGTM

File: test/api/views/test_contributions.py [hunk 27]
```diff
@@ -538,25 +535,55 @@ def test_update_contribution_type(self):
             description="Lorem ipsum dolor sit amet.",
             contribution_type_id="-KSDQcv0D",
         )
-        db.session.add(contrib)
-        db.session.commit()
-        old_contrib_contribution_type_id = Contribution.query.filter_by(id=contrib_id).first().contribution_type_id
-        data = {"contribution_type_id": "-KSDQcv9D"}
+        contrib.save()
+        old_contrib = (Contribution.query.filter_by(id=contrib_id).first()).serialize()
```
Comment (njirap) [inline test/api/views/test_contributions.py]: `Contribution.query.get(contrib_id)`

Comment (andela-milesanmi) [inline test/api/views/test_contributions.py]: done

File: api/views/contributions.py [hunk 28]
```diff
@@ -296,3 +338,4 @@ def patch(self, contribution_id):
             response = {"message": "Vote has been successfully registered"}, 201
         db.session.commit()
         return response
+
```
Comment (njirap) [inline api/views/contributions.py]: remove whitespace on last blank line

File: api/views/contributions.py [hunk 29]
```diff
@@ -141,34 +146,71 @@ def put(self, contrib_id):
...
+        return response
+            # response = jsonify(dict(
+            #     contribution=result,
+            #     status=200,
+            #     message='Contribution was successfully updated'))
+        # else:
+        #     response = jsonify(dict(status=404,
+        #                             message='Contribution does not exist'))
+        #     response.status_code = 404
+        #     return response
```
Comment (njirap) [inline api/views/contributions.py]: commented code here..

File: api/views/contributions.py (hunk 28 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: fixed

File: api/views/contributions.py (hunk 29 above)
Comment (andela-milesanmi) [inline api/views/contributions.py]: @njirap fixed.
